### ISSUE TITLE
Improve speed

### ISSUE THREAD

**Issue by deffi420** [resilar/crchack#2]
The CRC computation is too slow for files >100MB. This can be improved a lot.

![sonic](https://cloud.githubusercontent.com/assets/5674089/14366574/07279f68-fd1c-11e5-95e0-9f399b84b915.jpg)

**Comment by resilar** [comment]
We could use [Nayuki's method](https://www.nayuki.io/page/forcing-a-files-crc-to-any-value) with loss of generality. But first, let's try to speed up the CRC function with precomputed tables because the cost of forging is roughly _b_ times the computation of _CRC(msg)_, where _b_ is the number of modifiable bits, and _CRC(msg)_ is currently slow as hell.  We should also handle the case that _b_ is a lot larger than the CRC polynomial size _w_. After these improvements, the resulting cost is about _min(b,w) \* CRC(msg)_, which is acceptable if _CRC(msg)_ is reasonably fast.

**Comment by resilar** [comment]
The cost of forging has been reduced to **O**(n + b w³ *log* n) by exploiting **O**(w³ *log* n)-time sparse CRC calculation for b messages consisting of n-1 zeros and a single 1-bit.

---

The sparse CRC calculation utilizes (w**×**w)-bit *difference*-matrix **A** of differences (bit flips) caused by each input bit inversion in a w-bit window. **X** is solved from **AX**=**B** where **B** is the (w**×**w)-bit difference-matrix of the *next* w-bit window so that **AX**²=**C** represents the differences of the second w-bit window, and so on. Sparse CRC calculation precomputes *log* n difference-matrix squarings: **X**¹, (**X**¹)²=**X**², (**X**²)²=**X**⁴, (**X**⁴)²=**X**⁸  ... For a n-bit input, the difference caused by the *i*th bit (0 ≤ *i* < n) is obtained as the product of **O**(*log* n) multiplications of the precomputed (w**×**w)-bit differences-matrices.

In conclusion, the method constructs the final (b**×**w)-bit matrix for forging in **O**(b w³ *log* n) time assuming each (w**×**w)-bit matrix multiplication takes **O**(w³) time and the CRC of the input message is known.

---

**TLDR** Forging of CRC checksums is fast. For long messages, calculating the non-sparse CRC of the input message dominates the total running time in practice. This issue will be closed once the (non-sparse) CRC calculation has been optimized for speed as well.
